rat/parser.py

```diff
@@ -585,26 +611,99 @@ def expression(self, min_precedence=0, is_lhs=False):
+                self.remove()  # "match"
+                self.expect_token(Identifier)
+                bounded_variable = self.peek()
+                if bounded_variable.value == "rev":
+                    reverse_order = True
+                    self.remove()  # rev
+
+                    self.expect_token(Special, "(")
+                    self.remove()  # (
+                    self.expect_token(Identifier)  # variable name
+                    bounded_variable = self.peek()
+                    self.remove()  # variable
+                    self.expect_token(Special, ")")
+                    self.remove()  # )
+                else:
+                    self.remove()  # variable
+                self.expect_token(Special, "{")
+                self.remove()  # {
+                # we are now inside the match statement body
+                initial_declarations = {}
+                recurrence_equation = None
+                while True:
+                    next_token = self.peek()
+                    if next_token.value == "}":
+                        self.remove()  # }
+                        break
+                    if next_token.value == "first":
+                        self.remove()  # "first"
+                        self.expect_token(Operator, "=")
+                        self.remove()  # =
+                        rhs = self.expression()
+                        self.expect_token(Terminate)
+                        self.remove()  # ;
+                        initial_declarations["first"] = rhs
+                    elif next_token.value == "else":
+                        self.remove()  # "else"
+                        self.expect_token(Operator, "=")
+                        self.remove()  # =
+                        rhs = self.expression()
+                        self.expect_token(Terminate)
+                        self.remove()  # ;
+                        recurrence_equation = rhs
+                    else:
+                        raise ParseError(
+                            f"Found unknown keyword {next_token.value} within match statement",
+                            self.model_string,
+                            next_token.line_index,
+                            next_token.column_index,
+                        )
+
+                if not recurrence_equation:
+                    raise ParseError(
+                        "Match statement missing else(recurrence equation) stataement!", self.model_string, lhs.line_index, lhs.column_index
+                    )
+                return_statement = MatchedStatement(
+                    lhs=lhs,
+                    initial_declarations=initial_declarations,
+                    recurrence_equation=recurrence_equation,
+                    bounded_variable_name=bounded_variable.value,
+                    reverse_order=reverse_order,
+                    line_index=lhs.line_index,
+                    column_index=lhs.column_index,
```

Comment:
Excellent excellent I believe this is all the info necessary

Reply:
I'm not sure what keywords other than `first` would be needed for the initial declarations.